### PR TITLE
Sync

### PR DESCRIPTION
**Pull request by ulzar**
This implements the file transfer api functionalities.

### CODE REVIEW

File: athera/sync/client.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import grpc
+from athera.sync.sirius.services import service_pb2
+from athera.sync.sirius.services import service_pb2_grpc
+import sys
+import io 
+
+MAX_CHUNK_SIZE = 1024 * 1024  # 1Mb
+
+class Client(object):
+    # ToDo: CloseConnection for every method
```
Comment (sipickles) [inline athera/sync/client.py]: TODOs....

File: athera/sync/client.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        try:
+            mountsResponse = self.stub.Mounts(request, metadata=metadata)
+            return mountsResponse.mounts, None
+        except grpc.RpcError as e:
+            logging.debug("grpc.RpcError %s", e)
+            return None, e
+        except AttributeError as e:
+            return None, e
+
+    def get_files(self, group_id, mount_id, path="/"): # ToDo: User provide a buffer
```
Comment (sipickles) [inline athera/sync/client.py]: Move TODO into docstring and explain further

File: athera/sync/client.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                destination_file.write(resp.bytes)
+            logging.debug("Successfully wrote {} bytes into {}".format(total_bytes, destination_file.name))
+        except grpc.RpcError as e:
+            return e
+        except AttributeError as e:
+            return e
+
+    def upload_to_file(self, group_id, mount_id, file_to_upload, destination_path, chunk_size=MAX_CHUNK_SIZE):
+        """
+        Upload a file by chunks of up to 1 Mb
```
Comment (sipickles) [inline athera/sync/client.py]: Punctuation needed

File: test/api/test_storage.py [hunk 4]
```diff
@@ -88,36 +88,34 @@ def test_create_delete_gcs_driver(self):
         self.assertEqual(mount["type"], "MountTypeGroupCustom")
         self.assertEqual(mount["name"], fake_name)
 
-    # def test_rescan_driver(self):
-    #     """ Positive test - List the mounts the authenticated user has in this group """
-    #     driver_id = environment.ATHERA_API_TEST_HOME_DRIVER_ID
-    #     status = self.get_driver_indexing_status(driver_id)
-    #     self.assertEqual(status, False)
+    def test_rescan_driver(self):
+        """ Positive test - List the mounts the authenticated user has in this group """
```
Comment (sipickles) [inline test/api/test_storage.py]: docstring is wrong

File: test/api/test_storage.py [hunk 5]
```diff
@@ -88,36 +88,34 @@ def test_create_delete_gcs_driver(self):
...
+        self.assertEqual(response.status_code, codes.ok)
         
 
-    # def test_dropcache_driver(self):
-    #     """ Positive test - List the mounts the authenticated user has in this group """
-    #     driver_id = environment.ATHERA_API_TEST_ORG_DRIVER_ID
-    #     status = self.get_driver_indexing_status(driver_id)
-    #     self.assertEqual(status, False)
+    def test_dropcache_driver(self):
+        """ Positive test - List the mounts the authenticated user has in this group """
```
Comment (sipickles) [inline test/api/test_storage.py]: docstring is wrong

File: test/assets/hello.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+hey there.
```
Comment (sipickles) [inline test/assets/hello.txt]: Change to:
"Hello Atherian!"

File: test/sync/test_client.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    def setUpClass(cls):
+        cls.token = environment.ATHERA_API_TEST_TOKEN
+        url = environment.ATHERA_API_TEST_SYNC_URL
+        if url == "":
+            url = "localhost:9001"
+        cls.client = Client(url, cls.token)
+
+    def test_get_mounts(self):
+        """ Test we can get groups for the authenticated user.
+        Will only work with Python 2. Python 3 will segfault - TODO check this still applies
```
Comment (sipickles) [inline test/sync/test_client.py]: check this?

File: test/sync/test_client.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            )
+
+        for _, err in filesGenerator:
+            self.assertIsNone(err, "Got unexpected error: {}".format(err))
+
+    def test_download_file(self):
+        """ Test we can get files for the authenticated user.
+        Will only work with Python 2. Python 3 will segfault - TODO check this still applies
+        """
+        donwload_path = "downloaded_file.txt"
```
Comment (sipickles) [inline test/sync/test_client.py]: download_path

File: athera/sync/client.py (hunk 1 above)
Comment (sipickles) [inline athera/sync/client.py]: TODOs....

File: athera/sync/client.py (hunk 2 above)
Comment (sipickles) [inline athera/sync/client.py]: Move TODO into docstring and explain further

File: athera/sync/client.py (hunk 3 above)
Comment (sipickles) [inline athera/sync/client.py]: Punctuation needed

File: athera/sync/client.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            for resp in response:
+                total_bytes += resp.bytes_remaining
+                destination_file.write(resp.bytes)
+            logging.debug("Successfully wrote {} bytes into {}".format(total_bytes, destination_file.name))
+        except grpc.RpcError as e:
+            return e
+        except AttributeError as e:
+            return e
+
+    def upload_to_file(self, group_id, mount_id, file_to_upload, destination_path, chunk_size=MAX_CHUNK_SIZE):
```
Comment (sipickles) [inline athera/sync/client.py]: def upload_file

File: athera/sync/client.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                    ('active-group', group_id)]
+
+        try:
+            response = self.stub.FilesList(request, metadata=metadata, timeout=10)
+            for resp in response:
+                yield resp, None
+        except grpc.RpcError as e:
+            yield None, e
+
+    def download_to_file(self, group_id, mount_id, destination_file, path="/", chunk_size=MAX_CHUNK_SIZE): # ToDo: User provide a buffer
```
Comment (sipickles) [inline athera/sync/client.py]: def download_file

File: athera/sync/client.py (hunk 9 above)
Comment (sipickles) [inline athera/sync/client.py]: Provide primary upload region argument. Explain in docs:

_'region' specifies the ingress point for the data. Use the region closest to you. Other regions may have to perform a 'rescan' on the mount_id to detect the newly uploaded file._

File: athera/sync/client.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        Download a file by chunks of up to 1 Mb
```
Comment (sipickles) [inline athera/sync/client.py]: explain args more, eg:

'destination_file' is an absolute path on your local machine to save the file to. [CHECK IF THIS IS JUST THE PATH OR WHOLE FILENAME]

File: athera/sync/client.py (hunk 9 above)
Comment (sipickles) [inline athera/sync/client.py]: Explain args more, eg:

file_to_upload: an absolute path to a file on the local machine
destination_path: The target location to write the file on Athera, relative to the root of the mount.

File: athera/sync/client.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    # ToDo: CloseConnection for every method
+    # ToDo: Before performing any api call, check if token is expired; if so: REFRESH IT
+    """
+    Client to query the remote grpc file sync service, Sirius.
+    """
+
+    def __init__(self, url, token):
+        
+        self.url = url
+        # self.chunk_size = chunk_size
```
Comment (sipickles) [inline athera/sync/client.py]: remove

File: athera/sync/client.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+import logging
+import grpc
+from athera.sync.sirius.services import service_pb2
+from athera.sync.sirius.services import service_pb2_grpc
+import sys
+import io 
+
+MAX_CHUNK_SIZE = 1024 * 1024  # 1Mb
+
+REGION_NAMES = {
```
Comment (owenhaynes) [inline athera/sync/client.py]: Bad name is more then just region names?

File: athera/sync/client.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        Upload a file by chunks of up to 1 Mb.
+        mount_id allows you to select on which mount you would like to upload your file.
+        file_to_upload is a file object of the file to upload, read access is enough.
+        destination_path is the path on the mount at which the file will be uploaded (relative to the mount root).
+            Let's say the root path of the mount (you want to upload your file to) is /data/org/default-my-org
+            And within this mount, you want to upload your file in the folder 'uploads'
+            Let's say your filename is 'my_secret_movie.mov'
+            Your 'destination_path' would be: 'uploads/my_secret_movie.mov'
+        """
+        if chunk_size > MAX_CHUNK_SIZE:
```
Comment (owenhaynes) [inline athera/sync/client.py]: should we not return a error if chunk size is too big.

File: test/sync/__init__.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import os
+import sys
+# Allow imports from project root
+current_dir = os.path.dirname(os.path.realpath(__file__))
+root_dir = os.path.abspath(os.path.join(current_dir, ".."))
+
+if root_dir not in sys.path:
+    sys.path.append(root_dir)
```
Comment (owenhaynes) [inline test/sync/__init__.py]: no new line at the end of file

File: test/api/test_sessions.py [hunk 16]
```diff
@@ -177,7 +183,7 @@ def test_start_session_bad_payload(self):
         response = sessions.start_session(
             environment.ATHERA_API_TEST_BASE_URL,
             environment.ATHERA_API_TEST_GROUP_ID,
-            environment.ATHERA_API_TEST_TOKEN,
+            self.token,
             payload,
         )
         self.assertEqual(response.status_code, codes.bad_request)
```
Comment (owenhaynes) [inline test/api/test_sessions.py]: no new line at end of file

File: test/api/test_groups.py [hunk 17]
```diff
@@ -134,7 +140,7 @@ def test_get_group_users_wrong_target(self):
         response = groups.get_group_users(
             environment.ATHERA_API_TEST_BASE_URL,
             environment.ATHERA_API_TEST_GROUP_ID,
-            environment.ATHERA_API_TEST_TOKEN,
+            self.token,
             environment.ATHERA_API_TEST_OTHER_GROUP_ID,
         )
         self.assertEqual(response.status_code, codes.not_found)
```
Comment (owenhaynes) [inline test/api/test_groups.py]: no new line at end of ifile

File: athera/sync/client.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+                destination_file.write(resp.bytes)
+            logging.debug("Successfully wrote {} bytes into {}".format(total_bytes, destination_file.name))
+        except grpc.RpcError as e:
+            return e
+        except AttributeError as e:
+            return e
+
+    def upload_file(self, group_id, mount_id, file_to_upload, destination_path, chunk_size=MAX_CHUNK_SIZE):
+        """
+        Upload a file by chunks of up to 1 Mb.
```
Comment (sipickles) [inline athera/sync/client.py]: Tone of comment is too informal (too much use of 'second person'), and arguments format is inconsistent with other functions. How about this:

       """
        Upload a file by chunks of up to 1 Mb.

        mount_id: Storage Mount to upload file to.
        file_to_upload: The file object of the file to upload, read access is enough.
        destination_path: The path on the mount where the file will be uploaded (relative to the mount root).

        An example:
        * The final location needs to be '/data/org/default-my-org/uploads/movie1.mov'.
        * The mount provides the root path of '/data/org/default-my-org'.
        * Therefore, the 'destination_path' would need to be 'uploads/my_secret_movie.mov'.
        """

File: athera/sync/client.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        mount_id allows you to select on which mount you would like to upload your file.
+        file_to_upload is a file object of the file to upload, read access is enough.
+        destination_path is the path on the mount at which the file will be uploaded (relative to the mount root).
+            Let's say the root path of the mount (you want to upload your file to) is /data/org/default-my-org
+            And within this mount, you want to upload your file in the folder 'uploads'
+            Let's say your filename is 'my_secret_movie.mov'
+            Your 'destination_path' would be: 'uploads/my_secret_movie.mov'
+        """
+        if chunk_size >= MAX_CHUNK_SIZE:
+            raise ValueError("chunk_size is too high, the maximum value for chunck_size is {} (1Mb)".format(MAX_CHUNK_SIZE))
```
Comment (sipickles) [inline athera/sync/client.py]: Please Yves. I can't spend all night asking you to fix capitalisation and spelling. Use a spell check. Chunck?

File: test/api/test_groups.py [hunk 20]
```diff
@@ -4,14 +4,20 @@
...
+        cls.token = os.getenv("ATHERA_API_TEST_TOKEN")
+        if not cls.token:
+            raise ValueError("ATHERA_API_TEST_TOKEN environment variable must be set")
+
     def test_get_orgs(self):
         """ Positive test """
         response = groups.get_orgs(
             environment.ATHERA_API_TEST_BASE_URL,
-            environment.ATHERA_API_TEST_TOKEN,
+            self.token,
```
Comment (sipickles) [inline test/api/test_groups.py]: Whats the point of this change?

**Comment by sipickles** [comment]
Will create a PR for the docstring updates, since I am being the pedant here :)

**Pull request merged**